core/src/main/scala/caliban/rendering/Renderer.scala

```diff
@@ -0,0 +1,559 @@
+    directivesRenderer.unsafeRender(directives, None, writer)
+    writer += ' '
+    writer += '{'
+    fields.foreach { field =>
+      fieldDefinitionRenderer.unsafeRender(field, Some(1), writer)
+    }
+    writer += '\n'
+    writer += '}'
+    writer += '\n'
+  }
+
+  private lazy val directiveRenderer: Renderer[Directive] = new Renderer[Directive] {
+    override def unsafeRender(d: Directive, indent: Option[Int], writer: StringBuilder): Unit = {
+      writer ++= " @"
+      writer ++= d.name
+      var first = true
+      d.arguments.foreach { case (name, value) =>
+        if (first) {
+          writer += '('
+          writer ++= name
+          writer ++= ": "
+          writer ++= value.toInputString
+          first = false
+        } else {
+          writer ++= ", "
+          writer ++= name
+          writer ++= ": "
+          writer ++= value.toInputString
+        }
+      }
+      if (!first) writer += ')'
+    }
+  }
+
+  private lazy val directivesRenderer: Renderer[List[Directive]] =
+    directiveRenderer.list
+
+  private def unsafeRenderDescription(
+    description: Option[String],
+    indent: Option[Int],
+    writer: StringBuilder,
+    newline: Boolean = true
+  ): Unit = {
+    val tripleQuote = "\"\"\""
+
+    def nlOrSp  = if (newline) writer += '\n' else writer += ' '
+    def nlOrNot = if (newline) writer += '\n' else writer
+
+    description.foreach {
+      case value if value.contains('\n') =>
+        val valueEscaped = value.replace(tripleQuote, s"\\$tripleQuote")
+        // check if it ends in quote but it is already escaped
+        if (value.endsWith("\\\"")) {
+          writer ++= tripleQuote
+          nlOrNot
+          writer ++= valueEscaped
+          nlOrNot
+          writer ++= tripleQuote
+        } else if (value.last == '"') {
+          writer ++= tripleQuote
+          nlOrNot
+          writer ++= valueEscaped
+          nlOrSp
+          writer ++= tripleQuote
+          // check if it ends in quote. We need to break the sequence of 4 or more '"'
+        } else {
+          // ok. No quotes at the end of value
+          writer ++= tripleQuote
+          nlOrNot
+          writer ++= valueEscaped
+          nlOrNot
+          writer ++= tripleQuote
+        }
+      case value                         =>
+        if (newline) pad(indent, writer)
+        unsafeFastString(value, writer)
+        nlOrSp
+    }
+  }
+
+  private[caliban] lazy val fieldDefinitionRenderer: Renderer[FieldDefinition] = new Renderer[FieldDefinition] {
+    override def unsafeRender(definition: FieldDefinition, indent: Option[Int], builder: StringBuilder): Unit =
+      definition match {
+        case FieldDefinition(description, name, arguments, tpe, directives) =>
+          builder += '\n'
+          unsafeRenderDescription(description, indent, builder)
+          pad(indent, builder)
+          builder ++= name
+          unsafeRenderArguments(arguments, builder)
+          builder ++= s": $tpe"
```

Comment:
Nit: I think this will be slightly more performant
```suggestion
          builder ++= s": "
          builder ++= tpe.toString
```

Reply:
You're right I missed this bit